### PR TITLE
Correctly bootstrap numpy in setup.py

### PR DESCRIPTION
**Pull request by andreabedini**
Correctly bootstrapping numpy in setup.py is practically impossible but I am trying to follow what is done in pandas. Their approach is mostly based on http://stackoverflow.com/a/21621689/104099.

### CODE REVIEW

**Comment by andreabedini** [comment]
If we put 'numpy' in setup_requirements, pip will try to install it before running the build. As explained on SO there's a small drawback

> Note, that using setup_requires has a subtle downside: numpy will not only be compiled before building extensions, but also when doing python setup.py --help, for example. 

This might be solved by changing setup_requirements depending on the command line paramenters but it might not be worth doing.

**Comment by andreabedini** [comment]
This should definitely solve #413, as you can run `pip install` without pre-installing any dependency.

**Comment by andreabedini** [comment]
The tests on Travis are only partially useful because in `.travis.yml` we do `pip install -r requirements.txt` prior to install pytables. If you want to test this yourself you should create a clean virtual environment and install pytables from source with `pip install .`

**Comment by andreabedini** [comment]
@FrancescAlted still useful for 3.2?

File: setup.py [hunk 1]
```diff
@@ -594,7 +601,7 @@ def get_cython_extfiles(extnames):
 
 setuptools_kwargs['extras_require'] = {}
 
-setuptools_kwargs['setup_requires'] = requirements
+setuptools_kwargs['setup_requires'] = [ r for r in requirements if 'numpy' in r ]
```
Comment (andreabedini) [inline setup.py]: This means numpy whatever version. Is there a better way to express this?

**Comment by andreabedini** [comment]
Inspired by http://pytest.org/latest/goodpractises.html#integration-with-setuptools-test-commands I think I solve the problem of `setup.py` unconditionally installing numpy which happens when using `setup_requires` now numpy is only installed when running `build_ext`.

**Comment by FrancescAlted** [comment]
Wow, looks pretty cool (and somewhat complex, but that is the fate of installation procedures in Python) And yes, we are totally on-time for next release.  Merging.

**Comment by FrancescAlted** [comment]
Uh, it does not seem to work for me:

```
$ pip install .
<snip>
  File "/tmp/pip-5Vz4O1-build/setup.py", line 50, in run

    numpy_incl = pkg_resources.resource_filename('numpy', 'core/include')

  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 953, in resource_filename

    return get_provider(package_or_requirement).get_resource_filename(

  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 233, in get_provider

    __import__(moduleOrReq)

ImportError: No module named numpy

----------------------------------------
Cleaning up...
Command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-5Vz4O1-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-8JYr8X-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip-5Vz4O1-build
Storing debug log for failure in /home/faltet/.pip/pip.log
```

**Comment by FrancescAlted** [comment]
Ups, false alarm.  After more retries, it seems to work as intended :)

**Comment by andreabedini** [comment]
Thanks @FrancescAlted, do you have any clue on what caused the problem?

**Comment by FrancescAlted** [comment]
No idea, but I can reproduce the correct behavior, so hopefully nothing to worry about.

**Comment by andreabedini** [comment]
Reviewed files:
- setup.py @ r4

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/pytables/pytables/439)

**Comment by karlacio** [comment]
Hi,

I'm pretty new to PyTables and I hope my problem belongs to here.

I'm trying to install (with pip) PyTables on my mac.
I already have numpy installed but at some point during installation setup.py download numpy sources and try to compile it.
From setup.py output:
    Requirement already up-to-date: numpy>=1.7.1 in /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from tables)

but after copying some files:
    running build_ext
    Searching for numpy>=1.7.1
    Reading http://pypi.python.org/simple/numpy/
    Best match: numpy 1.9.2
    Downloading https://pypi.python.org/packages/source/n/numpy/numpy-1.9.2.zip#md5=e80c19d2fb25af576460bb7dac

Is setup.py doing something wrong or it's just some other problem?

Thanks,
Carlo.

**Comment by andreabedini** [comment]
Hi Carlo, thanks for trying out PyTables and thanks for reporting this. This is interesting, in this pull request I have indeed introduced a mechanism that install numpy on `build_ext` but it shouldn't install it again if it's already available. Moreover I just tested in a virtualenv with python 2.7.9 and pip 6.1.1 and I couldn't reproduce this behaviour (I also tried pip 1.5.6). I have installed numpy first with `pip install numpy` and tables next `pip install tables`. I have few questions: 1) are you installing pytables using pip or running `setup.py` ? 2) does the installation eventually succeed?

**Comment by karlacio** [comment]
Hi Andrea,

some software versions I'm using:
OSX 10.7.5
pip 6.1.1
python 2.7.2

and the replies:
1) On first attempt I tried to install PyTables with pip and I got the problem I described on my previous post. Then I dowloaded PyTables 3.2.0 sources and tried to compile it with the command:

    python setup.py build_ext --inplace --hdf5 /cdvadded/hdf5

in both cases it looks like that, after:
  *running build_ext
   Searching for numpy>=1.7.1*
it assumes that numpy isn't installed.

2) in both cases the installation failed.

I have output of both installations attempts but I think they are too big to be included here.
Can I put them somewhere?

Carlo.

**Comment by andreabedini** [comment]
You can upload the log to gist.guthub.com and link it here.

**Comment by andreabedini** [comment]
gist.github.com (sorry I'm on my phone)

**Comment by karlacio** [comment]
Here they are:

- [pip log] (https://gist.github.com/karlacio/334b22fc3508e394e8f8)
- [setup.py log] (https://gist.github.com/karlacio/56a60e5ca5823659f528)

hope I have linked them correctly.

**Comment by karlacio** [comment]
Update:

I should have something of wrong on my machine.
I tried to install just tables inside a virtualenv and it worked.

Sorry for the trouble.

**Comment by andreabedini** [comment]
> `SystemError: Cannot compile 'Python.h'. Perhaps you need to install python-dev|python-devel.`
definitely indicates some other problem. A virtualenv does help.

**Comment by karlacio** [comment]
Looking around to solve my problem (on just one machine) I saw that also someone else with other packages had problems with setup trying to re-install numpy while it's already installed.

Even from my setup logs it looks like that PyTables setup tries to re-compile numpy and then fails.

Just a curiosity: why it tries to compile/install numpy even if it's there and of a compatible version?
Thanks

**Comment by andreabedini** [comment]
@karlacio the `setup.py` shipped in 3.2.0 includes [this](https://github.com/PyTables/PyTables/blob/develop/setup.py#L55) line that installs numpy if it's not already there. It is not supposed to installed it again. I later realised that, with recent versions of `pip`, that line is not even needed and I plan to remove it soon.

If it causes problems to just, please install numpy manually and delete line 55 of `setup.py`.

PS: perhaps we should stop chatting on this PR and move this discussion to https://gitter.im/PyTables/PyTables.

**Comment by karlacio** [comment]
Sure, thanks again!

**Pull request merged**